### PR TITLE
fix: enable discount code checks for over consumption

### LINKED ISSUE #5865: Able to book more discounted or access restricted tickets than allowed
Even if the organizer only allows 1 Discount Ticket. The discount ticket code still works after 1 ticket was sold. The same applies for access codes.

Expected: Discount Tickets and Access Codes should only work for the number of tickets specified.

E.g.:
* if the user chooses to buy two tickets, only one ticket should be discounted, if only one discount ticket is left.
* if all discount tickets are already sold out, the discount option should not show up on the ticket page. Instead there should be a message "discount tickets sold out" when a discount ticket code is entered in the UI.

Screenshot from 2019-05-03 16-51-37

### LINKED ISSUE #6102: sub issue : prevent order placement on potential overuse of discount code
Parent Issue : #5865 

This issue prevents the order to be placed in the case the discount codes will be potentially overused.

### PR DESCRIPTION
**Pull request by uds5501**
Fixes #6102  

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [ ] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:
This PR fixes the existing tests for checking if Discount Code is exhausted or not upon application

### CODE REVIEW

**Comment by uds5501** [comment]
@mrsaicharan1 @shreyanshdwivedi please review

**Comment by uds5501** [comment]
@mrsaicharan1 @shreyanshdwivedi @prateekj117 Take a look?

**Comment by uds5501** [comment]
@mrsaicharan1 @shreyanshdwivedi @prateekj117 Please review

File: app/api/helpers/ticketing.py [hunk 1]
```diff
@@ -24,16 +25,16 @@ def get_order_expiry():
...
-
+        ticket_ids = [ticket.id for ticket in discount_code.tickets]
+        old_holders = get_count(TicketHolder.query.filter(TicketHolder.ticket_id.in_(ticket_ids))
+                                .join(Order).filter(Order.status.in_(['completed', 'placed'])))
         for holder in ticket_holders:
             ticket_holder = TicketHolder.query.filter_by(id=holder).one()
-            if ticket_holder.ticket.id in discount_code.tickets.split(","):
+            if ticket_holder.ticket.id in ticket_ids:
                 qty += 1
         if (qty + old_holders) <= discount_code.tickets_number and \
```
Comment (mrsaicharan1) [inline app/api/helpers/ticketing.py]: At this condition, what you could do is, suppose the maximum number of tickets on which the discount code can be applied is 3 and you are trying to order 5 tickets, apply the discount only on 3. This is how it's implemented on other platforms.

Comment (mrsaicharan1) [inline app/api/helpers/ticketing.py]: At this condition, what you could do is, suppose the maximum number of tickets on which the discount code can be applied is 3 and you are trying to order 5 tickets, apply the discount only on 3. This is how it's implemented on other platforms.

Comment (uds5501) [reply app/api/helpers/ticketing.py]: @mrsaicharan1 should I break this into 2 sub-issues?

- This PR could serve as a check for discount code's over consumption 
while the second sub issue will focus on application of discount code in the way you suggested?

Comment (mrsaicharan1) [reply app/api/helpers/ticketing.py]: @uds5501 Yeah, that would be better

Comment (uds5501) [reply app/api/helpers/ticketing.py]: @mrsaicharan1 I have linked it to the appropriate sub-issue. Please take a look

**Comment by uds5501** [comment]
@iamareebjamal Understood, will open an issue regarding this.

**Comment by mrsaicharan1** [comment]
@uds5501 We can merge this and work on DB level queries in a follow-up PR as soon as you resolve the conflicts :D

**Pull request merged**